src/Database/Kioku/Internal/TrieIndex.hs

```diff
@@ -173,26 +176,38 @@ trieLookupMany keys (TI buf rootDrop root) =
             | otherwise =
               cont
 
+          -- | Gets a list containing the results for all the multikeys in the argument for the remainingArc
+          goKids :: [MultiKey] -> [Int]
           goKids [] = rest
           goKids (kid:restKids) =
             case breakCommonPrefix (mkPrefix kid) remainingArc of
               (_, "", "") ->
-                readValues 0 (goKids restKids)
-              (_, remainingKid, "") ->
+                readValues 0 (goKids restKids) -- get all the values for this index, prepended to the remaining kid's results
+              (_, remainingKid, "") -> -- if there is a remainder for the prefix and the arc was completely consumed, we descend into the subtries
                 trySubtries (kid { mkPrefix = remainingKid })
                             0
                             (goKids restKids)
 
               _ -> goKids restKids
 
+          -- | Reads out all the values from the index and prepend them to a list of other values
+          readValues :: Int -> [Int] -> [Int]
           readValues n cont
             | n < valueCount = readValueN n : readValues (n + 1) cont
             | otherwise = cont
 
       in case (remainingKey, remainingArc) of
-         ("", "") -> readValues 0 (goKids kids)
-         (_, "") -> trySubtries (MultiKey remainingKey kids) 0 rest
-         _ -> rest
+           -- if both the subkey and the arc were consumed, we will branch on the children of the MultiKey
+           ("", "") -> readValues 0 (goKids kids)
+           -- if the key was consumed but there was some remaining arc, we try prepending the subkey to prefix
+           -- of each child in the MultiKey and performing a search with each of those.
+           ("", _) ->
+             let kids' = map (\(MultiKey prefix children) -> MultiKey (subkey <> prefix) children) kids
+              in foldl' (\r k -> go k arcDrop offset r) rest kids'
```

Comment:
I geuss `foldl'` is my default because it's more efficient memory wise. If we want laziness then I think writing a tail recursive function may be best.

If by rescanning the involved keys you mean the subkey that has already been consumed, we could address this by adding the length of the subkey to the arcDrop argument of `go` instead of prepending subkey to each of the MultiKey children (since we already know it matches). That way no work is  being duplicated.

Reply:
Ok, I understand now.